app-notifier/src/main/kotlin/org/ostelco/prime/appnotifier/FirebaseAppNotifier.kt

```diff
@@ -45,6 +45,24 @@ class FirebaseAppNotifier: AppNotifier {
                             body = FCMStrings.JUMIO_IDENTITY_FAILED.s,
                             data = data)
                 }
+                NotificationType.PAYMENT_METHOD_REQUIRED -> {
+                    logger.info("Notifying customer $customerId of failed payment on renewal of " +
+                            "subscription to product ${data["sku"]} with data $data")
+                    sendMessage(customerId = customerId,
+                            title = FCMStrings.SUBSCRIPTION_RENEWAL_TITLE.s,
+                            body = FCMStrings.SUBSCRIPTION_PAYMENT_METHOD_REQUIRED.s,
+                            data = data)
+                }
+                NotificationType.USER_ACTION_REQUIRED -> {
+                    /* TODO: Add support for 3D secure notification. */
+                }
+                NotificationType.SUBSCRIPTION_RENEWAL_UPCOMING -> {
+
+                }
+                NotificationType.SUBSCRIPTION_RENEWAL_STARTING -> {
+                    /* No notification, as a notification will be sent on either successful
+                       or failed renewal. */
+                }
             }
```

Comment:
This API is a bit messed up.  `enum NotificationType` can have fields such as `title` and `body` which is then directly used instead of using `when` cases.  Also, if FCMStrings is specific to Firebase, they should not be in generic API interfaces - `prime-modules`.

The simplest way to design it is to think of `AppNotifier` as an interface whereas `FirebaseAppNotifier` as one of the many possible implementations of `AppNotifier`.
We should be able to replace whole FCM with some other implementation with just adding and removing dependency in the gradle build file and no need of any code change.

Reply:
https://www.notion.so/redotter/Review-FCM-delivery-setup-reliability-and-message-format-0701c7da03e24a89963cc7bdd4f7aa76